
This is an attempt at an equivalent, portable version of:

http://anonscm.debian.org/viewvc/kernel/dists/trunk/linux/debian/patches/bugfix/all/radeon-firmware-is-required-for-drm-and-kms-on-r600-onward.patch?revision=20909&view=co

--- a/sys/dev/drm2/radeon/r600.c
+++ b/sys/dev/drm2/radeon/r600.c
@@ -3012,6 +3012,12 @@
 		rdev->accel_working = false;
 	}
 
+	/* Don't start up if the ucode is missing. */
+	if (!rdev->me_fw || !rdev->pfp_fw || !rdev->rlc_fw) {
+		DRM_ERROR("radeon: ucode required for R600+.\n");
+		return -EINVAL;
+	}
+
 	return 0;
 }
 
--- a/sys/dev/drm2/radeon/rv770.c
+++ b/sys/dev/drm2/radeon/rv770.c
@@ -1186,6 +1186,12 @@
 		rdev->accel_working = false;
 	}
 
+	/* Don't start up if the ucode is missing. */
+	if (!rdev->me_fw || !rdev->pfp_fw || !rdev->rlc_fw) {
+		DRM_ERROR("radeon: ucode required for R600+.\n");
+		return -EINVAL;
+	}
+
 	return 0;
 }
 
